nilearn/connectome/connectivity_matrices.py

```diff
@@ -231,6 +231,77 @@ def sym_to_vec(symmetric, discard_diagonal=False):
     return symmetric[..., tril_mask] / scaling[tril_mask]
 
 
+def vec_to_sym_matrix(vec, diagonal=None):
+    """Return the symmetric matrix given its flattened lower triangular part.
+    Acts on the last dimension of the array if not 1-dimensional.
+    Diagonal can be encompassed in vec or given separately. In both cases, note
+    that diagonal elements are multiplied by sqrt(2).
+
+    Parameters
+    ----------
+    vec : numpy.ndarray, shape (..., p * (p + 1) /2)
+        The input array.
+
+    diagonal : numpy.ndarray, shape (..., p), optional
+        The diagonal array to be stacked to vec.
+
+    Returns
+    -------
+    sym : numpy.ndarray, shape (..., p, p) or (..., p + 1, p + 1) if diagonal
+        is given seperately.
+        The output symmetric array.
+
+    Note
+    ----
+    This function is meant to be the inverse of sym_matrix_to_vec.
+
+    See also
+    --------
+    nilearn.connectome.sym_matrix_to_vec
+    """
+    n = vec.shape[-1]
+    # solve p * (p + 1) / 2 = n subj. to p > 0
+    # p ** 2 + p - 2n = 0 & p > 0
+    # p = - 1 / 2 + sqrt( 1 + 8 * n) / 2
+    p = (sqrt(8 * n + 1) - 1.) / 2
+    if p > floor(p):
+        raise ValueError(
+            "Vector of unsuitable shape {0} can not be transformed to "
+            "a symmetric matrix.".format(vec.shape))
+
+    p = int(p)
+    first_shape = vec.shape[:-1]
+    if diagonal is None:
+        n_columns = p
+        diagonal_offset = 0
+    else:
+        if diagonal.shape[:-1] != first_shape or diagonal.shape[-1] != p + 1:
```

Comment:
>diagonal should be given

`vec_to_sym_matrix` is meant to be the reverse to `sym_matrix_to_vec`. So if you have discarded the diagonal in `sym_matrix_to_vec`, you need to provide it separately to reconstruct the symmetric matrix. For instance this can be useful for correlation matrices for which we know the diagonal is 1.

> should be same as length of features or samples 
if you provide it, it should have the right shape, that is for n-dimensional array: same as `vec` until n-1, and matching the matrix shape on the last dimension